hpx/hpx_init_impl.hpp

```diff
@@ -297,56 +387,73 @@ namespace hpx
-            main_f = util::bind_back(detail::init_helper, f);
-        std::vector<std::string> cfg;
         util::function_nonser<void()> const empty;
 
         HPX_ASSERT(argc != 0 && argv != nullptr);
 
-        return init(main_f, desc_commandline, argc, argv, cfg,
-            empty, empty, mode);
+        hpx::init_params iparams;
+        iparams.f = util::bind_back(detail::init_helper, f);
+        iparams.desc_cmdline_ptr = std::make_shared<options_description>(
+            desc_commandline);
+        iparams.argc = argc;
+        iparams.argv = argv;
+        iparams.startup = empty;
+        iparams.shutdown = empty;
+        iparams.mode = mode;
+        return init(iparams);
     }
 
     // Main entry point for launching the HPX runtime system.
     inline int
     init(util::function_nonser<int(int, char**)> const& f,
         int argc, char** argv, hpx::runtime_mode mode)
     {
-        std::string app_name(HPX_APPLICATION_STRING);
-        return init(f, app_name, argc, argv, mode);
+        util::function_nonser<void()> const empty;
+
+        HPX_ASSERT(argc != 0 && argv != nullptr);
+
+        hpx::init_params iparams;
+        iparams.f = util::bind_back(detail::init_helper, f);
+        iparams.argc = argc;
+        iparams.argv = argv;
+        iparams.startup = empty;
+        iparams.shutdown = empty;
+        iparams.mode = mode;
+        return init(iparams);
     }
 
     inline int
     init(util::function_nonser<int(int, char**)> const& f,
         int argc, char** argv, std::vector<std::string> const& cfg,
         hpx::runtime_mode mode)
     {
-        std::string app_name(HPX_APPLICATION_STRING);
-        using hpx::program_options::options_description;
-
-        options_description desc_commandline(
-            "Usage: " + app_name +  " [options]");
-
-        util::function_nonser<int(hpx::program_options::variables_map& vm)>
-            main_f = util::bind_back(detail::init_helper, f);
-
         HPX_ASSERT(argc != 0 && argv != nullptr);
 
-        return init(main_f, desc_commandline, argc, argv, cfg,
-            startup_function_type(), shutdown_function_type(), mode);
+        hpx::init_params iparams;
+        iparams.f = util::bind_back(detail::init_helper, f);
+        iparams.argc = argc;
+        iparams.argv = argv;
+        iparams.cfg = cfg;
+        iparams.mode = mode;
+        return init(iparams);
     }
 
     inline int
         init(util::function_nonser<int(int, char**)> const& f,
         std::vector<std::string> const& cfg,
         hpx::runtime_mode mode)
     {
-        char *dummy_argv[2] = { const_cast<char*>(HPX_APPLICATION_STRING), nullptr };
+        util::function_nonser<void()> const empty;
 
-        return init(f, 1, dummy_argv, cfg, mode);
+        hpx::init_params iparams;
+        iparams.f = util::bind_back(detail::init_helper, f);
+        iparams.cfg = cfg;
+        iparams.startup = empty;
+        iparams.shutdown = empty;
```

Comment:
Aren't these empty by default? Otherwise you can also do
```suggestion
        iparams.shutdown = {};
```

Reply:
For now, the default value is `startup_function_type()`, I'll do this change !